<div class="alert alert-block alert-success">
    
## Data Science Libraries Quiz

### Instructions


1. Write code which will produce the specified result.  
2. Enter that result in the green response cell. 
3. Write comment(s) to describe the code 


To receive full credit, make sure you have at least 1 cell with code and an answer in the green response cell. 



#### Reminders

1. Complete the work without the help of others.
2. You can receive partial credit for your code.





---
<div class="alert alert-block alert-success">

### Rubric (50 points total)


#### Dataframes Exercises (8)
##### Total points = 5
| Description |  Points | 
| ---- | ---- |
| No attempt | 0   |
| Correct result | 2  |
| Appropriate functions | 1 
| Appropriate syntax | 1 
| Code comments | 1 | 

#### Visualization Exercises (2)

##### Total points = 5

| Description |  Points | 
| ---- | ---- |
| No attempt | 0   |
| Plotted data | 2  |
| Axis labels | 1 | 
| Titles | 1 |
| Customization | 1 | 

<div class="alert alert-block alert-info">

### Setup
 
1. Run the import statements
2. Load the datset `'tips'` from pydataset.
3. Assign this to the dataframe `df`.

In [7]:
# Import statements
from pydataset import data
import pandas as pd
import matplotlib.pyplot as plt
import math

In [2]:
# Load the dataframe from pydataset
# Assign to df 
df = data('tips')

---

<div class="alert alert-block alert-info">
  
### Question 1
    
        
a) What is the highest bill?
    
b) What time of day did this bill occur? 

In [3]:
df

total_bill   tip     sex smoker   day    time  size
1         16.99  1.01  Female     No   Sun  Dinner     2
2         10.34  1.66    Male     No   Sun  Dinner     3
3         21.01  3.50    Male     No   Sun  Dinner     3
4         23.68  3.31    Male     No   Sun  Dinner     2
5         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
240       29.03  5.92    Male     No   Sat  Dinner     3
241       27.18  2.00  Female    Yes   Sat  Dinner     2
242       22.67  2.00    Male    Yes   Sat  Dinner     2
243       17.82  1.75    Male     No   Sat  Dinner     2
244       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [4]:
df.sort_values(by = 'total_bill').tail(1)

total_bill   tip   sex smoker  day    time  size
171       50.81  10.0  Male    Yes  Sat  Dinner     3

<div class="alert alert-block alert-success">

### Question 1 Response
    
a) $50.81
    
b) Dinner


---

<div class="alert alert-block alert-info">
  
### Question 2
    
    
a) What is the most common table size?

b) Are the majority of all tables smoking or non-smoking?

AttributeError: 'DataFrameGroupBy' object has no attribute 'math'

AttributeError: 'numpy.int64' object has no attribute 'mode'

<div class="alert alert-block alert-success">

### Question 2 Response
    
a)
    
b)


---

<div class="alert alert-block alert-info">
  
### Question 3
    
    
a) What proportion of the bills are from the weekend (Saturday and/or Sunday)?

Note: Provide your answer rounded to two decimal places.

<div class="alert alert-block alert-success">

### Question 3 Response
    
a)



---
<div class="alert alert-block alert-info">
    
## Question 4

-  Create a column named `percent_tip`.


- Calculate the percent_tip based on the formula below. 



$\texttt{Percent tip} = \dfrac{\texttt{tip amount}}{\texttt{bill}}$

a) What was the highest percent tip?

<div class="alert alert-block alert-success">

### Question 4 Response
    
a)

---

<div class="alert alert-block alert-info">
    
## Question 5

    
    
a) Which time of day has a higher average bill, dinner or lunch?
    
b) What is the difference in the average bill?


<div class="alert alert-block alert-success">

### Question 5 Response
    
a)

b)

---
<div class="alert alert-block alert-info">
    
## Question 6


a) What was the total revenue for the restaurant (bill and tips)

b) What percentage of the revenue came from tips?


<div class="alert alert-block alert-success">

### Question 6 Response
    
a)

b)

---
<div class="alert alert-block alert-info">
    
## Question 7

The following code produces errors. 
    
Correct it to determine the **maximum bill** for dinner time on Thursday
    
    
    
a) Describe one error in the code.
    
b) What is the maximum bill for dinner time on Thursday?




#### Incorrect code
```python
df.total_bill( (df.day == 'Thursday') and (df.time == 'dinner') ) 
```

In [ ]:
#### Correct the code below
#### Run it to find the maximum bill for dinner time on Thursday

df.total_bill((df.day == 'Thursday') and (df.time == 'dinner')) 

In [18]:
df.total_bill [ (df.day == 'Thurs') and (df.time == 'Dinner') ] 

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

<div class="alert alert-block alert-success">

### Question 7 Response
    
a) Describe one error in the code.
    The first set of paratheses used should be square brackets
    
b) What is the maximum bill for dinner time on Thursday?

In [ ]:
df

In [15]:
df

total_bill   tip     sex smoker   day    time  size
1         16.99  1.01  Female     No   Sun  Dinner     2
2         10.34  1.66    Male     No   Sun  Dinner     3
3         21.01  3.50    Male     No   Sun  Dinner     3
4         23.68  3.31    Male     No   Sun  Dinner     2
5         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
240       29.03  5.92    Male     No   Sat  Dinner     3
241       27.18  2.00  Female    Yes   Sat  Dinner     2
242       22.67  2.00    Male    Yes   Sat  Dinner     2
243       17.82  1.75    Male     No   Sat  Dinner     2
244       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

---
<div class="alert alert-block alert-info">
    
## Question 8
    

a) Rename the column named `size` to `table_size`. 


Make sure this column name replaces the former column name. 


<div class="alert alert-block alert-success">

### Question 8 Response

    
a) List the column names of the dataframe below
    
    

---
<div class="alert alert-block alert-info">
    
## Question 9

- Use matplotlib to create a **scatter plot** for total_bill and tips. 
    - Plot total_bill along the x-axis

- Customize the visualization in at least one way.

- Add descriptive axis labels and titles.


<div class="alert alert-block alert-success">
    
### Question 9 Response



<div class="alert alert-block alert-info">
    
## Question 10

- Use matplotlib to create a **histogram** for total bill

- Customize the visualization in at least one way.

- Add descriptive axis labels and titles.


<div class="alert alert-block alert-success">

### Question 10 Response